bestie/server/testresult.go

```diff
@@ -0,0 +1,220 @@
+	bes "github.com/enfabrica/enkit/third_party/bazel/buildeventstream" // Allows prototext to automatically decode embedded messages
+	"google.golang.org/genproto/googleapis/devtools/build/v1"
+	"google.golang.org/protobuf/proto"
+)
+
+type bigQueryTable struct {
+	project   string
+	dataset   string
+	tablename string
+}
+
+type keyValuePair struct {
+	key   string
+	value string
+}
+
+type testMetric struct {
+	metricname string
+	tags       []keyValuePair
+	value      float64
+	timestamp  int64
+}
+
+type metricTestResult struct {
+	table   bigQueryTable
+	metrics []testMetric
+}
+
+// Pertinent fields to uniquely identify a Bazel event stream.
+type bazelStream struct {
+	buildId       string
+	invocationId  string
+	testName      string
+	run           string
+	invocationSha string // derived
+}
+
+// Derive a unique invocation SHA value.
+func deriveInvocationSha(items []string) string {
+	// The SHA value allows combining one or more items
+	// (e.g. invocationId, buildId, run) to uniquely identify
+	// a bazel test stream without impacting the database schema.
+	hash := sha256.Sum256([]byte(strings.Join(items[:], ".")))
+	return hex.EncodeToString(hash[:])
+}
+
+// Store fields that help identify this Bazel stream.
+func identifyStream(bazelBuildEvent bes.BuildEvent, streamId *build.StreamId) bazelStream {
+	// Extract the stream identifier fields of interest.
+	stream := bazelStream{
+		buildId:      streamId.GetBuildId(),
+		invocationId: streamId.GetInvocationId(),
+		run:          strconv.Itoa(int(bazelBuildEvent.GetId().GetTestResult().GetRun())),
+		testName:     bazelBuildEvent.GetId().GetTestResult().GetLabel(),
+	}
+	// Calculate a SHA256 hash using the following fields to uniquely identify this stream.
+	stream.invocationSha = deriveInvocationSha([]string{stream.invocationId, stream.buildId, stream.run})
+	return stream
+}
+
+// Handle metrics extraction from the TestResult event.
+func handleTestResultEvent(bazelBuildEvent bes.BuildEvent, streamId *build.StreamId) error {
+	stream := identifyStream(bazelBuildEvent, streamId)
+	m := bazelBuildEvent.GetTestResult()
+	if m == nil {
+		return fmt.Errorf("Error extracting TestResult data from event message")
+	}
+
+	fmt.Printf("TestResult for %s: %s\n", stream.testName, m.GetStatus())
+	fmt.Printf("  run: %s\n", stream.run)
+	fmt.Printf("  buildId: %s\n", stream.buildId)
+	fmt.Printf("  invocationId: %s\n", stream.invocationId)
+	fmt.Printf("  invocationSha: %s\n\n", stream.invocationSha)
+
+	outputFiles := m.GetTestActionOutput()
+	var ofname, ofuri string
+	found := false
+	for _, of := range outputFiles {
+		ofname = of.GetName()
+		ofuri = of.GetUri()
+		if strings.HasSuffix(ofname, "outputs.zip") {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil
+	}
+
+	// Strip off any file:// prefix from the URI to access the local file system path.
+	filePrefix := "file://"
+	if strings.HasPrefix(ofuri, filePrefix) {
+		ofuri = ofuri[len(filePrefix):]
+	}
```

Comment:
This almost certainly won't work when we put this up into Cloud Run, because:
* remote build URLs are going to not start with `file://`, but this code will fail to strip the URL prefix, and then try to load the file on-disk
* local builds file URLs won't make sense to the BES endpoint, which is on another system and therefore doesn't have the same filesystem

...but we can fix that later.

Reply:
We'll need to address the Cloud Run case later. I could see how the URL was formatted for local builds, but wasn't sure about how this would be formatted for remote builds. That's why I added the HasPrefix check to cover the local case, knowing there will likely be some similar adjustment needed for the remote case. Definitely need to revisit this later on.